qa/L0_trace/opentelemetry_unittest.py

```diff
@@ -153,121 +198,560 @@ def _check_events(self, span_name, events):
-    input1_data = np.full(shape=(1, 16), fill_value=-1, dtype=np.int32)
+            len(span_names),
+            sum(expected_counts.values()),
+            "Unexpeced number of span names collected",
+        )
+        for name, count in expected_counts.items():
+            self.assertEqual(
+                span_names.count(name),
+                count,
+                "Unexpeced number of " + name + " spans collected",
+            )
 
-    inputs.append(client.InferInput("INPUT0", [1, 16], "INT32"))
-    inputs.append(client.InferInput("INPUT1", [1, 16], "INT32"))
+    def _verify_nesting(self, spans, expected_parent_span_dict):
+        """
+        Helper function that checks parent-child relationships between
+        collected spans are the same as in `expected_parent_span_dict`.
+
+        Args:
+            spans (List[dict]): list of json objects, extracted from the trace and
+                   containing span info. For this test `name`
+                   and `events` are required.
+            expected_parent_span_dict (dict): dictionary, containing expected
+                   parents and children in the dictionary form:
+                        <parent_span_name> (str) : <children_names> (List[str])
+        """
+        seen_spans = {}
+        for span in spans:
+            cur_span = span[0]["spanId"]
+            seen_spans[cur_span] = span[0]["name"]
+
+        parent_child_dict = {}
+        for span in spans:
+            cur_parent = span[0]["parentSpanId"]
+            cur_span = span[0]["name"]
+            if cur_parent in seen_spans.keys():
+                parent_name = seen_spans[cur_parent]
+                if parent_name not in parent_child_dict:
+                    parent_child_dict[parent_name] = []
+                parent_child_dict[parent_name].append(cur_span)
+
+        for key in parent_child_dict.keys():
+            parent_child_dict[key].sort()
+
+        self.assertDictEqual(parent_child_dict, expected_parent_span_dict)
+
+    def _verify_headers_propagated_from_client_if_any(self, root_span, headers):
+        """
+        Helper function that checks traceparent's ids, passed in clients
+        headers/metadata was picked up on the server side.
+        If `headers` are None, checks that `root_span` does not have
+        `parentSpanId` specified.
+
+        Args:
+            root_span (List[dict]): a json objects, extracted from the trace and
+                   containing root span info. For this test `traceID`
+                   and `parentSpanId` are required.
+            expected_parent_span_dict (dict): dictionary, containing expected
+                   parents and children in the dictionary form:
+                        <parent_span_name> (str) : <children_names> (List[str])
+        """
+        parent_span_id = NO_PARENT_SPAN_ID
+
+        if headers != None:
+            parent_span_id = headers["traceparent"].split("-")[2]
```

Comment:
So context propagation is making sure that the headers associated with the request are included in the trace?

Reply:
I am not sure if `making sure` is the right word. 

Context propagation describes a set of operations that needs to be performed on client and server side that results in a unified trace. The steps are:
On client side, headers/metadata needs to be populated with `traceId` and `spanId` of parent's trace (in a special format) , there is more info, that gets injected, but I'l omit it for now.

On server side, these headers should be extracted, and started trace should have the same `traceId` as passed, and newly started spans should have parent span with `id`, passed in headers.

If nothing found in headers/metadata, we start a fresh new trace with new `traceId`

Reply:
Got it, thanks for correcting my understanding.